packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -32,12 +32,23 @@ export interface GraphQLRequestListener<TContext = Record<string, any>> {
       'document' | 'operationName' | 'operation'
     >,
   ): ValueOrPromise<void>;
+  // If this hook is defined, it is invoked immediately before GraphQL execution
+  // would take place. If its return value resolves to a non-null
+  // GraphQLResponse, that result is used instead of executing the query. It is
+  // an error for more than one plugin to return a value that resolves to a
+  // non-null GraphQLResponse.
+  execute?(
```

Comment:
I agree with the idea of adding a hook like this, but the name `execute` seems confusing. This isn't really execution, but rather a way to short-circuit execution. We don't invoke `executionDidStart` when a response is returned from cache for example, because we don't measure execution time in that case. Maybe something like `responseForOperation`?

Reply:
There is also possible confusion/overlap with the addition of a pluggable executor, which was added in https://github.com/apollographql/apollo-server/commit/185dd45f4360d10fb50801b7fbb26df98db4d19b and is used to plug in Apollo Gateway. I think we want to keep this separate, both because the gateway isn't a regular plugin (it composes a schema from a service list and it needs to reload itself when service list changes) and because we actually want to be able to cache gateway responses.